Sources/PackageCollections/PackageCollections.swift

```diff
@@ -100,12 +101,11 @@ public struct PackageCollections: PackageCollectionsProtocol {
                 if sources.isEmpty {
                     return callback(.success([]))
                 }
-                let lock = Lock()
-                var refreshResults = [Result<Model.Collection, Error>]()
+                let refreshResults = ThreadSafeArrayStore<Result<Model.Collection, Error>>()
                 sources.forEach { source in
                     self.refreshCollectionFromSource(source: source) { refreshResult in
-                        lock.withLock { refreshResults.append(refreshResult) }
-                        if refreshResults.count == (lock.withLock { sources.count }) {
```

Comment:
this was the issue that prompted the PR. I saw a few PRs failing on PackageCollectionsTests.testHappyRefresh inconsistently so I was hunting for a race condition (interesting TSAN did not report this). the issue here is that the lock is on `sources` instead of `refreshResults`. so in addition to fixing this, I decided to create this new utility for arrays so we dont have to mess with locks

Reply:
I fixed this coding mistake in https://github.com/apple/swift-package-manager/pull/3108 though didn't anticipate it causing race condition. Anyway, 👍